### PR TITLE
EZP-27458: Add Field Facet Visitor

### PR DESCRIPTION
**Pull request by wizhippo**
Add filed facet support

Example 
```
           new FacetBuilder\FieldFacetBuilder([
                'name' => 'Specialties',
                'fieldPaths' =>['agent/specialties/tag_ids'],
                'limit' => 9999,
            ]),
           
           new FacetBuilder\FieldFacetBuilder([
                'name' => 'Multipath',
                'fieldPaths' => [
                     'article/title',
                     'atrilcle/tags/tag_ids',
                ],
                'limit' => 9999,
           ]),
```

Related PR https://github.com/ezsystems/ezpublish-kernel/pull/2557

### CODE REVIEW

**Comment by andrerom** [comment]
We would somehow need to make sure we have integration tests in kernel, but besides that LGTM

File: lib/Query/Common/FacetBuilderVisitor/Field.php [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    public function visitBuilder(FacetBuilder $facetBuilder, $fieldId)
+    {
+        $parameters = [];
+        $criteria = new Criterion\MatchAll();
+        $fieldPaths = $facetBuilder->fieldPaths;
+
+        if (count($fieldPaths) > 1) {
+            $contentTypeIdentifier = array_shift($fieldPaths);
```
[the remark this answers is not in the document]
Comment (andrerom) [reply lib/Query/Common/FacetBuilderVisitor/Field.php]: > Though I see we have some integration tests, which use FieldFacetBuilder, in \eZ\Publish\API\Repository\Tests\SearchServiceTest so I wonder how do they work (maybe there were skipped, or just don't throw when missing builder?).

> AFAICS $fieldPaths contains for e.g. array('article/title'), so this doesn't add up. Of course existing usage in integration tests might be wrong.

hmm, I might need to dive into these too. I did _some_ clean up when adding the current three facetbuilders.

[the remark this answers is not in the document]
Comment (andrerom) [reply lib/Query/Common/FacetBuilderVisitor/Field.php]: So as this is first implementation, we can consider changing the design. Haven't looked at this myself in depth yet, but do you remember we change the design for this?

[the remark this answers is not in the document]
Comment (wizhippo) [reply lib/Query/Common/FacetBuilderVisitor/Field.php]: I think I have found my disconnect   It should be an array of values like 'fieldPaths' => array('article/title'), 

I find it a bit odd that you can specify multiple fieldPaths.  I guess that would act the same as adding multiple  builders?

Just to clarify then the syntax should be 

new FacetBuilder\FieldFacetBuilder([
    'name' => 'Multipath',
    'fieldPaths' => [
        'article/title',
        'atrilcle/tags/tag_ids',
    ],
    'limit' => 9999,
]),

[the remark this answers is not in the document]
Comment (wizhippo) [reply lib/Query/Common/FacetBuilderVisitor/Field.php]: I have updated to use the 'article/title' format.

File: lib/Query/Common/FacetBuilderVisitor/Field.php [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    private function getSort(FieldFacetBuilder $facetBuilder)
+    {
+        switch ($facetBuilder->sort) {
+            case FieldFacetBuilder::COUNT_DESC:
+                return 'count';
+            case FieldFacetBuilder::TERM_ASC:
+                return 'index';
+        }
+
+        return 'index';
```
Comment (andrerom) [inline lib/Query/Common/FacetBuilderVisitor/Field.php]: Should throw NotImplemented on anything else to make it clear FieldFacetBuilder::COUNT_ASC and FieldFacetBuilder::TERM_DESC is not supported here.

File: lib/Query/Common/FacetBuilderVisitor/Field.php [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+    public function visitBuilder(FacetBuilder $facetBuilder, $fieldId)
+    {
+        $parameters = [];
+
+        foreach ($facetBuilder->fieldPaths as $fieldPath) {
+            $parts = explode('/', $fieldPath);
+
+            if (count($parts) > 1) {
+                $contentTypeIdentifier = array_shift($parts);
```
Comment (andrerom) [inline lib/Query/Common/FacetBuilderVisitor/Field.php]: Not used anywhere, can be `new Criterion\ContentTypeIdentifier(array_shift($parts));`

File: lib/Query/Common/FacetBuilderVisitor/Field.php [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+                $parameters = array_merge($parameters, [
+                    'facet.field' => "{!ex=dt key={$fieldId}}{$fieldName}",
+                    "f.{$fieldName}.facet.limit" => $facetBuilder->limit,
+                    "f.{$fieldName}.facet.mincount" => $facetBuilder->minCount,
+                    "f.{$fieldName}.facet.sort" => $this->getSort($facetBuilder),
+                    "f.{$fieldName}.facet.missing" => 'true',
+                ]);
+                if ($facetBuilder->prefix) {
+                    array_merge($parameters, [
+                        "f.{$fieldName}.facet.prefix" => '"' . $this->escapeQuote($facetBuilder->prefix, true) . '"',
```
Comment (wizhippo) [inline lib/Query/Common/FacetBuilderVisitor/Field.php]: Not sure if escaping is needed here?  Any solr users can confirm?

File: lib/Query/Common/FacetBuilderVisitor/Field.php [hunk 5]
```diff
@@ -83,20 +84,16 @@ public function visitBuilder(FacetBuilder $facetBuilder, $fieldId)
...
             if (count($parts) > 1) {
-                $contentTypeIdentifier = array_shift($parts);
-                $criteria = new Criterion\ContentTypeIdentifier($contentTypeIdentifier);
+                $criteria = new Criterion\ContentTypeIdentifier(array_shift($parts));
             } else {
                 $criteria = new Criterion\MatchAll();
             }
 
-            $fieldDefinitionIdentifier = array_shift($parts);
-            $name = array_shift($parts);
```
Comment (andrerom) [inline lib/Query/Common/FacetBuilderVisitor/Field.php]: In this case it's different, without the variable name it's not possible to know what those values are unless you know by hand signature of getFieldTypes. `ContentTypeIdentifier` is kind of self explanatory.

**Comment by wizhippo** [comment]
Any chance the v3 is coming up we can visit this and facets?

**Comment by alongosz** [comment]
> Any chance the v3 is coming up we can visit this and facets?

ping @SylvainGuittard

**Comment by SylvainGuittard** [comment]
@alongosz I am not against having this in v3.
This PR is against master. Is this branch compatible with Symfony 4 and current eZ Platform v3? Is there any other blocker?

**Comment by andrerom** [comment]
>  Is there any other blocker?

Someone in eZ Eng needs to write integration tests in kernel, besides that we are more or less good to go here.

**Comment by wizhippo** [comment]
Superseded by https://github.com/ezsystems/ezplatform-kernel/pull/94

**Pull request closed without merging**